File: system/metrics.py
"""
system/metrics.py
DIX VISION v42.2 — Prometheus Metrics

Tracks: trade_latency_ms, hazard_detection_time_ms,
        governance_decision_time_ms, circuit_breaker_triggers.

# Backends

This module has two interchangeable backends and picks one at import
time:

* **Rust (``dixvision_py_system``)** — PyO3 extension built from
  ``rust/py_system/``. Preferred when the wheel is installed. Counters
  + histograms live in Rust behind a single ``parking_lot::Mutex``;
  the FFI surface takes plain primitives on every call. Histograms
  are stored as ``Vec<f64>`` with the same 10_000 → 5_000 ring-buffer
  trim as the Python reference.
* **Pure Python (fallback)** — the original threading-lock
  implementation kept verbatim so pre-polyglot ledgers still replay
  on a box without the Rust wheel.

Both backends satisfy the same invariants, validated by
``tests/test_metrics_parity.py``:

* ``increment(name, value, labels)`` bumps a counter. Labels are
  folded into the counter key as ``"{name}:{labels}"`` when non-empty.
* ``observe(name, value_ms)`` appends a histogram sample with the
  same 10_000 → 5_000 ring-buffer trim on both sides.
* ``p99(name)`` is the 99th-percentile sample across the current
  buffer, matching the Python reference's indexing scheme
  (``int(len * 0.99)`` clamped to ``len - 1``).
* ``snapshot()`` returns ``{"counters": {..}, "p99": {..}}`` as a
  plain ``dict`` with identical keys on both sides.

Selection is observable via ``backend()``. Both implementations are
importable directly via ``make_python_sink()`` and ``make_rust_sink()``
so parity tests can exercise them without reloading this module or
mutating the environment.
"""
from __future__ import annotations

import threading
import time
from collections import defaultdict
from typing import Optional

# ---------------------------------------------------------------- Rust backend
# Probe the Rust wheel at import time. The wheel is optional; when it
# is missing we fall back to the Python implementation below without
# raising. This keeps `pytest` green on boxes without cargo.
try:
    import dixvision_py_system as _rs  # type: ignore[import-not-found]

    _HAVE_RUST = all(
        hasattr(_rs, fn)
        for fn in (
            "metrics_increment",
            "metrics_observe",
            "metrics_p99",
            "metrics_snapshot",
        )
    )
except ImportError:  # pragma: no cover — covered by the no-wheel path
    _rs = None  # type: ignore[assignment]
    _HAVE_RUST = False


# -------------------------------------------------------------- Python backend


class _PythonMetricsSink:
    """Pure-CPython reference sink.

    Preserves the exact public API of the pre-polyglot
    ``MetricsSink`` class so callers importing ``MetricsSink`` from
    this module keep working when no Rust wheel is available.
    """

    __slots__ = ("_lock", "_counters", "_histograms")

    _HISTOGRAM_CAP = 10_000
    _HISTOGRAM_KEEP = 5_000

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._counters: dict[str, float] = defaultdict(float)
        self._histograms: dict[str, list[float]] = defaultdict(list)

    def increment(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[dict] = None,
    ) -> None:
        key = _compose_key(name, labels)
        with self._lock:
            self._counters[key] += value

    def observe(self, name: str, value_ms: float) -> None:
        with self._lock:
            buf = self._histograms[name]
            buf.append(value_ms)
            if len(buf) > self._HISTOGRAM_CAP:
                self._histograms[name] = buf[-self._HISTOGRAM_KEEP :]

    def p99(self, name: str) -> float:
        with self._lock:
            vals = sorted(self._histograms.get(name, []))
            if not vals:
                return 0.0
            idx = int(len(vals) * 0.99)
            return vals[min(idx, len(vals) - 1)]

    def snapshot(self) -> dict:
        with self._lock:
            # Read p99 inline under the same lock so the snapshot is
            # internally consistent even if writers keep going.
            counters = dict(self._counters)
            p99 = {}
            for name, buf in self._histograms.items():
                if not buf:
                    p99[name] = 0.0
                    continue
                sorted_buf = sorted(buf)
                idx = int(len(sorted_buf) * 0.99)
                p99[name] = sorted_buf[min(idx, len(sorted_buf) - 1)]
            return {"counters": counters, "p99": p99}


def _compose_key(name: str, labels: Optional[object]) -> str:
    """Match the Rust seam exactly: empty / ``None`` labels collapse
    to the bare name; otherwise the key becomes ``"{name}:{labels}"``."""
    if labels is None:
        return name
    return f"{name}:{labels}" if labels else name


# ---------------------------------------------------------------- Rust wrapper


class _RustMetricsSink:
    """Thin adapter delegating to the process-global Rust sink.

    All instances share state because the FFI seam exposes a single
    ``OnceLock<MetricsSink>``. That mirrors the Python module-level
    ``get_metrics()`` singleton so migration is transparent. Tests
    that want isolation must use :class:`_PythonMetricsSink` directly.
    """

    __slots__ = ("_rs",)

    def __init__(self, rust_module: object) -> None:
        self._rs = rust_module

    def increment(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[dict] = None,
    ) -> None:
        # Preserve the Python reference's key format by pre-folding
        # labels into the name on the Python side. The Rust seam does
        # the same fold, but only when the caller passes a non-empty
        # label string; routing it here keeps one code path.
        if labels is None:
            self._rs.metrics_increment(name, float(value), None)
            return
        label_str = "" if not labels else str(labels)
        self._rs.metrics_increment(name, float(value), label_str or None)

    def observe(self, name: str, value_ms: float) -> None:
        self._rs.metrics_observe(name, float(value_ms))

    def p99(self, name: str) -> float:
        return float(self._rs.metrics_p99(name))

    def snapshot(self) -> dict:
        counters_list, p99_list = self._rs.metrics_snapshot()
        return {
            "counters": {k: float(v) for k, v in counters_list},
            "p99": {k: float(v) for k, v in p99_list},
        }


# -------------------------------------------------------- Public compatibility


# Preserve the pre-polyglot public name. When the Rust wheel is
# available, ``MetricsSink()`` constructions point to the shared Rust
# sink; otherwise they construct a fresh Python sink. Callers using
# the canonical ``get_metrics()`` accessor below get the same
# behaviour on both paths — parallel ``MetricsSink()`` instantiations
# are discouraged but remain compatible with the Python reference.
def make_python_sink() -> _PythonMetricsSink:
    """Force-construct a fresh pure-Python sink. Used by the parity
    test suite; callers should prefer :func:`get_metrics`."""
    return _PythonMetricsSink()


def make_rust_sink() -> _RustMetricsSink:
    """Force-construct a Rust-backed wrapper. Raises ``RuntimeError``
    when the Rust wheel was not imported — tests gate on
    ``_rust_wheel_available()`` before calling this."""
    if _rs is None or not _HAVE_RUST:
        raise RuntimeError("dixvision_py_system wheel not available")
    return _RustMetricsSink(_rs)


def backend() -> str:
    """Return ``"rust"`` or ``"python"`` depending on which
    implementation backs the process-global sink. Observable for
    logging / health checks."""
    return "rust" if (_HAVE_RUST and _rs is not None) else "python"


MetricsSink = _RustMetricsSink if (_HAVE_RUST and _rs is not None) else _PythonMetricsSink


class LatencyTimer:
    """Context manager for measuring latency.

    Backend-agnostic — takes any sink object that exposes
    ``observe(name, value_ms)``.
    """

    __slots__ = ("_sink", "_name", "_start")

    def __init__(self, sink: object, metric_name: str) -> None:
        self._sink = sink
        self._name = metric_name
        self._start = 0.0

    def __enter__(self) -> LatencyTimer:
        self._start = time.perf_counter()
        return self

    def __exit__(self, *_) -> None:
        elapsed_ms = (time.perf_counter() - self._start) * 1000.0
        self._sink.observe(self._name, elapsed_ms)


# -------------------------------------------------------------- Singleton path


_metrics: object = None
_init_lock = threading.Lock()


def get_metrics() -> object:
    """Process-wide metrics sink. First call constructs the sink using
    whichever backend was selected at import time; subsequent calls
    return the same instance."""
    global _metrics
    if _metrics is None:
        with _init_lock:
            if _metrics is None:
                if _HAVE_RUST and _rs is not None:
                    _metrics = _RustMetricsSink(_rs)
                else:
                    _metrics = _PythonMetricsSink()
    return _metrics

File: rust/system/src/metrics.rs
//! # Metrics sink (T0-10 — runtime port)
//!
//! Port of the Python `MetricsSink` reference implementation in
//! `system/metrics.py`. Thread-safe in-memory counters + histograms
//! with the exact same public-API semantics so the parity suite
//! passes.
//!
//! * `increment(name, value, labels)` — bump a named counter. Labels
//!   are folded into the counter key as ``"{name}:{labels}"`` when
//!   non-empty, matching the Python code's key format.
//! * `observe(name, value_ms)` — append a histogram sample. When a
//!   metric exceeds [`HISTOGRAM_RING_CAP`] samples the tail
//!   [`HISTOGRAM_RING_KEEP`] are kept (same ring-buffer logic as the
//!   reference).
//! * `p99(name)` — 99th-percentile sample across the current buffer
//!   for `name`. Empty buffers report 0.0 (matches Python default).
//! * `snapshot()` — point-in-time copy of every counter and per-metric
//!   p99. Returned as owned primitives so callers can hand it across
//!   the FFI seam without borrowing from the sink.
//!
//! The Prometheus text-exposition format is explicitly out of scope
//! for this port; the manifest's T0-10 exporter will be added in a
//! follow-up PR. This crate only supplies the in-memory sink the
//! exporter will eventually read from.
//!
//! ## Concurrency
//!
//! Writers and readers both take a single [`parking_lot::Mutex`]. The
//! Python reference uses `threading.Lock` for identical purpose.
//! Tests exercise concurrent increment/observe/snapshot calls to
//! prove there are no torn reads.

use std::collections::HashMap;

use parking_lot::Mutex;

/// Upper bound on samples retained per histogram metric before the
/// ring-buffer drops the oldest half. Matches the Python constant
/// inlined in `MetricsSink.observe`.
const HISTOGRAM_RING_CAP: usize = 10_000;

/// Number of samples retained after a ring-buffer trim. Matches the
/// Python constant inlined in `MetricsSink.observe`.
const HISTOGRAM_RING_KEEP: usize = 5_000;

/// Percentile computed by [`MetricsSink::p99`]. 0.99 matches the
/// Python reference; if ever tuned, also update the name of the
/// accessor to preserve the public API contract.
const P99_QUANTILE: f64 = 0.99;

/// Read-only snapshot of every counter and per-metric p99.
///
/// Returned by [`MetricsSink::snapshot`] so callers can inspect state
/// without holding the sink lock. Keys are the same strings the
/// Python reference emits (counter keys may include a ``":labels"``
/// suffix).
#[derive(Debug, Clone, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct MetricsSnapshot {
    /// Per-counter totals, keyed by ``"{name}"`` or
    /// ``"{name}:{labels}"`` when labels were supplied at increment
    /// time.
    pub counters: HashMap<String, f64>,
    /// Per-histogram 99th-percentile values at the instant the
    /// snapshot was taken.
    pub p99: HashMap<String, f64>,
}

/// In-memory metrics sink.
///
/// Governance and Indira share one instance per process; see
/// [`crate::metrics`] for the canonical accessor via
/// `dixvision-py-system`. All operations acquire a single mutex —
/// the reference implementation does the same and the hot path is
/// well below the SLO budget even with the lock.
#[derive(Debug, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct MetricsSink {
    inner: Mutex<MetricsInner>,
}

#[derive(Debug, Default)]
struct MetricsInner {
    counters: HashMap<String, f64>,
    histograms: HashMap<String, Vec<f64>>,
}

impl MetricsSink {
    /// Construct an empty sink. Equivalent to `MetricsSink()` in the
    /// Python reference.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Bump a counter by `value`. When `labels` is non-empty the key
    /// is folded to ``"{name}:{labels}"`` to match the Python
    /// reference exactly; the caller is responsible for keeping
    /// `labels` stable across calls if they want aggregation.
    pub fn increment(&self, name: &str, value: f64, labels: Option<&str>) {
        let key = compose_key(name, labels);
        let mut guard = self.inner.lock();
        *guard.counters.entry(key).or_insert(0.0) += value;
    }

    /// Append a histogram sample (latency in milliseconds, by
    /// convention). Replicates the Python ring-buffer semantics.
    #[allow(clippy::significant_drop_tightening)]
    pub fn observe(&self, name: &str, value_ms: f64) {
        let mut guard = self.inner.lock();
        let buf = guard.histograms.entry(name.to_string()).or_default();
        buf.push(value_ms);
        if buf.len() > HISTOGRAM_RING_CAP {
            let excess = buf.len() - HISTOGRAM_RING_KEEP;
            buf.drain(..excess);
        }
    }

    /// 99th-percentile sample for `name`. Returns 0.0 when no samples
    /// have been observed, matching the Python reference which
    /// initialises the buffer to `[0.0]` inside `p99` but ignores
    /// that default when other samples exist.
    #[must_use]
    pub fn p99(&self, name: &str) -> f64 {
        let samples: Vec<f64> = {
            let guard = self.inner.lock();
            match guard.histograms.get(name) {
                Some(buf) if !buf.is_empty() => buf.clone(),
                _ => return 0.0,
            }
        };
        p99_of(&samples)
    }

    /// Point-in-time snapshot of every counter and per-metric p99.
    /// Safe to call concurrently with writers; the snapshot is
    /// taken while holding the lock, so readers never see torn
    /// state.
    #[must_use]
    pub fn snapshot(&self) -> MetricsSnapshot {
        let (counters, histograms) = {
            let guard = self.inner.lock();
            (guard.counters.clone(), guard.histograms.clone())
        };
        let p99 = histograms
            .iter()
            .map(|(k, buf)| {
                let v = if buf.is_empty() { 0.0 } else { p99_of(buf) };
                (k.clone(), v)
            })
            .collect();
        MetricsSnapshot { counters, p99 }
    }
}

fn compose_key(name: &str, labels: Option<&str>) -> String {
    match labels {
        Some(l) if !l.is_empty() => format!("{name}:{l}"),
        _ => name.to_string(),
    }
}

/// p99 helper. Computes the 99th-percentile of `buf` using the same
/// indexing scheme as the Python reference:
/// ``idx = int(len(vals) * 0.99); return sorted(vals)[min(idx, len-1)]``.
/// The buffer is cloned and sorted locally so the sink's stored
/// ordering is preserved.
#[allow(
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    clippy::cast_precision_loss,
    clippy::indexing_slicing
)]
fn p99_of(buf: &[f64]) -> f64 {
    let mut sorted: Vec<f64> = buf.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let len = sorted.len();
    // `len` is always ≥ 1 because callers check `is_empty` first.
    // The Python reference computes ``int(len * 0.99)`` then clamps
    // to ``len - 1``. Reproduce both steps so rounding behaviour is
    // identical.
    let idx = (len as f64 * P99_QUANTILE) as usize;
    let clamped = idx.min(len - 1);
    sorted[clamped]
}

#[cfg(test)]
#[allow(
    clippy::expect_used,
    clippy::unwrap_used,
    clippy::uninlined_format_args,
    clippy::float_cmp,
    clippy::cast_precision_loss
)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use std::thread;

    #[test]
    fn increment_without_labels_accumulates() {
        let s = MetricsSink::new();
        s.increment("trades", 1.0, None);
        s.increment("trades", 2.5, None);
        let snap = s.snapshot();
        assert_eq!(snap.counters.get("trades"), Some(&3.5));
    }

    #[test]
    fn increment_with_labels_keys_separately() {
        let s = MetricsSink::new();
        s.increment("trades", 1.0, Some("side=buy"));
        s.increment("trades", 2.0, Some("side=sell"));
        let snap = s.snapshot();
        assert_eq!(snap.counters.get("trades:side=buy"), Some(&1.0));
        assert_eq!(snap.counters.get("trades:side=sell"), Some(&2.0));
        assert!(!snap.counters.contains_key("trades"));
    }

    #[test]
    fn observe_reports_p99_from_buffer() {
        let s = MetricsSink::new();
        for v in 1..=100 {
            s.observe("latency_ms", f64::from(v));
        }
        // int(100 * 0.99) = 99, clamped to len-1=99, sorted[99] = 100.0
        assert_eq!(s.p99("latency_ms"), 100.0);
    }

    #[test]
    fn p99_of_empty_is_zero() {
        let s = MetricsSink::new();
        assert_eq!(s.p99("never_observed"), 0.0);
    }

    #[test]
    fn observe_ring_buffer_trims_on_overflow() {
        let s = MetricsSink::new();
        for v in 0..HISTOGRAM_RING_CAP + 10 {
            s.observe("noisy", v as f64);
        }
        // Buffer was trimmed when len exceeded CAP=10_000 to tail
        // KEEP=5_000. Subsequent 9 observes after the trim leave us
        // at KEEP + (CAP + 10 - CAP) - 1 = 5009 samples.
        let snap = s.snapshot();
        assert!(snap.p99.contains_key("noisy"));
    }

    #[test]
    fn snapshot_is_independent_of_sink() {
        let s = MetricsSink::new();
        s.increment("a", 1.0, None);
        let before = s.snapshot();
        s.increment("a", 100.0, None);
        // The earlier snapshot must not reflect the later increment.
        assert_eq!(before.counters.get("a"), Some(&1.0));
        assert_eq!(s.snapshot().counters.get("a"), Some(&101.0));
    }

    #[test]
    fn snapshot_reports_p99_per_metric() {
        let s = MetricsSink::new();
        s.observe("fast", 1.0);
        s.observe("slow", 1000.0);
        let snap = s.snapshot();
        assert_eq!(snap.p99.get("fast"), Some(&1.0));
        assert_eq!(snap.p99.get("slow"), Some(&1000.0));
    }

    #[test]
    fn concurrent_writers_do_not_lose_counts() {
        let sink = Arc::new(MetricsSink::new());
        let mut handles = Vec::new();
        for _ in 0..8 {
            let s = Arc::clone(&sink);
            handles.push(thread::spawn(move || {
                for _ in 0..1_000 {
                    s.increment("ctr", 1.0, None);
                }
            }));
        }
        for h in handles {
            h.join().expect("worker thread panicked");
        }
        let snap = sink.snapshot();
        assert_eq!(snap.counters.get("ctr"), Some(&8_000.0));
    }

    #[test]
    fn concurrent_snapshot_never_tears() {
        let sink = Arc::new(MetricsSink::new());
        let stop = Arc::new(parking_lot::Mutex::new(false));
        let writer = {
            let s = Arc::clone(&sink);
            let stop = Arc::clone(&stop);
            thread::spawn(move || loop {
                s.increment("w", 1.0, None);
                s.observe("lat", 2.0);
                if *stop.lock() {
                    break;
                }
            })
        };
        // Read a snapshot many times; the counters map must always
        // be an internally-consistent clone (all entries real
        // strings + finite f64s).
        for _ in 0..5_000 {
            let snap = sink.snapshot();
            for v in snap.counters.values() {
                assert!(v.is_finite());
            }
        }
        *stop.lock() = true;
        writer.join().expect("writer thread panicked");
    }
}
